xds/internal/balancer/cdsbalancer/cluster_handler.go

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cdsbalancer
+
+import (
+	"errors"
+	"sync"
+
+	"google.golang.org/grpc/internal/buffer"
+	xdsclient "google.golang.org/grpc/xds/internal/client"
+)
+
+
+// clusterHandler will be given a name representing a cluster. It will then update the CDS policy
+// constantly with a list of Clusters to pass down to XdsClusterResolverLoadBalancingPolicyConfig
+// in a stream like fashion.
+type clusterHandler struct {
+	// A mutex to protect entire tree of clusters.
+	clusterMutex sync.Mutex
+	root         *clusterNode
+
+	// A way to ping CDS Balanccer about any updates or errors to a Node in the tree.
+	// This will either get called from this handler constructing an update or from a child with an error.
+	updateChannel *buffer.Unbounded
```

Comment:
Is this the difference you mentioned in the comment? (channel instead of callback)
Either way is fine.

One thing to optimize is that, for this update, we only need to keep the most recent one, so this can be a `make(chan ..., 1)`. And when you need to send to the channel, do an unblock read to clear it first.

```go
select {
case <-updateCh:
default:
}
updateCh <- the-new-update
```

Reply:
Difference is that I kept mutex lock in handleresp callback rather than a higher level function in WatchCluster call.

Reply:
I based this off of the way xdsClient would write to updateChannel buffer.

Reply:
So just for conceptual understanding, there's two possible options here? Either giving the clusterhandler a callback with a reference to b.updateCh that it then writes to that memory, OR pass an explicit pointer to the same buffer for the clusterhandler to write directly to?

Reply:
I figured I would put an update on the buffer that run() in cds picks up directly.

Reply:
Changed the update channel to be length 1.